### PR TITLE
Prohibit taking implicit cross products with volatile operations

### LINKED ISSUE #1784: Inconsistency with volatile calls
1) `FOR x IN {1, 2, 3} UNION (x, random());` returns identical values
2) `FOR x IN {1, 2, 3} UNION (SELECT (x, random()));` returns different values
3) `SELECT (FOR x IN {1, 2, 3} UNION (SELECT (x, random())));` returns identical values (!)

I /think/ that all 3 ought to return different values but 2 and 3 definitely should do the same thing.

### PR DESCRIPTION
**Pull request by msullivan**
I wanted error spans on these messages, which meant threading around
a few more contexts in the compiler.

I needed to make the bodies of FOR scoped in order to have the
iterator variable recognized as having single cardinality inside the
body when the body is just an expression and not an explicit subquery.

Fixes #1784.

### CODE REVIEW

File: edb/edgeql/compiler/stmt.py [hunk 1]
```diff
@@ -181,14 +181,19 @@ def compile_ForQuery(
...
-            view_scls=ctx.view_scls,
-            view_rptr=ctx.view_rptr,
-            result_alias=qlstmt.result_alias,
-            view_name=ctx.toplevel_result_view_name,
-            forward_rptr=True,
-            ctx=sctx)
+        with sctx.newscope(fenced=True) as bctx:
+            stmt.result = setgen.scoped_set(
+                compile_result_clause(
+                    qlstmt.result,
```
Comment (elprans) [inline edb/edgeql/compiler/stmt.py:187]: `astutils.ensure_qlstmt(qlstmt.result)` here should be more efficient as it wouldn't introduce extraneous fences for expressions that are already statements.

File: edb/edgeql/compiler/inference/cardinality.py [hunk 2]
```diff
@@ -186,20 +188,49 @@ def _coalesce_cardinality(
...
+    # check this by assuming that a voltile operation is AT_MOST_ONE
+    # and making sure that the resulting cartesian cardinality isn't
+    # multi.
+    for i, vol in enumerate(vols):
+        if vol == VOLATILE:
+            cards2 = list(cards)
+            cards2[i] = AT_MOST_ONE
+            if cartesian_cardinality(cards2).is_multi():
+                raise errors.QueryError(
+                    "can not take cross product of volatile operation",
```
Comment (elprans) [inline edb/edgeql/compiler/inference/cardinality.py:212]: This also seems to trip up even if I put the volatile in `WITH`:

```
WITH rand := random() SELECT ({1, 2}, rand)
```

File: edb/edgeql/compiler/stmt.py (hunk 1 above)
Comment (msullivan) [reply edb/edgeql/compiler/stmt.py]: More efficient in what sense? Compiler efficiency, or will the generated code be improved?

`ensure_qlstmt` breaks a few tests---I haven't dug into it yet though.

Comment (elprans) [reply edb/edgeql/compiler/stmt.py]: I was mainly concerned with scope tree complexity, but the benefit here is negligible.  Let's keep your initial version then.  Technically, `ensure_qlstmt` should be equivalent to setting up a fence, but there's probably some weird interaction with iterator state.

**Review by elprans — APPROVED** [review]

**Pull request merged**